datumaro/plugins/open_images_format.py

```diff
@@ -349,38 +349,113 @@ def find_sources(cls, path):
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        if self._writer_manager:
+            self._writer_manager.__exit__(exc_type, exc_value, traceback)
+
+    def writerow(self, rowdict):
+        if not self._writer_manager:
+            self._writer_manager = self._writer_manager_factory()
+            self._writer = self._writer_manager.__enter__()
+
+        self._writer.writerow(rowdict)
+
+
+class _AnnotationWriter:
+    _POSSIBLE_ANNOTATION_FILE_PATTERNS = (
+        # class description files don't need to be listed,
+        # because they are always written
+        OpenImagesPath.FULL_IMAGE_DESCRIPTION_FILE_NAME,
+        *OpenImagesPath.SUBSET_IMAGE_DESCRIPTION_FILE_PATTERNS,
+        '*' + OpenImagesPath.LABEL_DESCRIPTION_FILE_SUFFIX,
+        '*' + OpenImagesPath.BBOX_DESCRIPTION_FILE_SUFFIX,
+        DEFAULT_IMAGE_META_FILE_NAME,
+    )
 
-        with open(absolute_path, 'w', encoding='utf-8', newline='') as f:
-            yield csv.DictWriter(f, field_names)
+    def __init__(self, root_dir):
+        self._annotations_dir = osp.join(root_dir, OpenImagesPath.ANNOTATIONS_DIR)
+        self._written_annotations = set()
 
-    def apply(self):
-        annotations_dir = osp.join(self._save_dir, OpenImagesPath.ANNOTATIONS_DIR)
+        os.makedirs(self._annotations_dir, exist_ok=True)
 
-        os.makedirs(annotations_dir, exist_ok=True)
+    def open(self, file_name, newline=None):
+        self._written_annotations.add(file_name)
+        f = open(
+            osp.join(self._annotations_dir, file_name),
+            'w', encoding='utf-8', newline=newline)
+        return f
 
-        self._save_categories()
-        self._save_label_category_parents()
-        self._save_subsets()
+    @contextlib.contextmanager
+    def open_csv(self, file_name, field_names, *, write_header=True):
+        with self.open(file_name, newline='') as f:
+            writer = csv.DictWriter(f, field_names)
+            if write_header:
+                writer.writeheader()
+            yield writer
+
+    def open_csv_lazy(self, file_name, field_names):
+        return _LazyCsvDictWriter(
+            lambda: self.open_csv(file_name, field_names))
+
+    def remove_unwritten(self):
+        for file_name in os.listdir(self._annotations_dir):
+            if file_name not in self._written_annotations and any(
+                fnmatch.fnmatch(file_name, pattern)
+                for pattern in self._POSSIBLE_ANNOTATION_FILE_PATTERNS
+            ):
+                os.unlink(osp.join(self._annotations_dir, file_name))
+
+class OpenImagesConverter(Converter):
+    DEFAULT_IMAGE_EXT = '.jpg'
 
-    def _save_categories(self):
-        with self._open_csv_annotation(
+    def apply(self):
+        self._save(_AnnotationWriter(self._save_dir))
+
+    @classmethod
+    def patch(cls, dataset, patch, save_dir, **options):
+        converter = cls(dataset, save_dir, **options)
```

Comment:
The purpose of patching is not only updating, but also avoiding to read the whole input dataset (which is supposed to be loaded lazily), if it is possible. Passing input dataset can be changed to passing only modified subsets in most cases. Consider doing it here as well.

Reply:
Unfortunately, that doesn't seem to be possible, because building the updated `images.meta` file requires all subsets. Should I switch to using per-subset meta files?

Reply:
Answered below